src/threadsync.cpp

```diff
@@ -461,14 +469,14 @@ bool ThreadSync::wrapperExecutionLockLockExcl()
 void ThreadSync::wrapperExecutionLockUnlock()
 {
   int saved_errno = errno;
-  if (WorkerState::currentState() != WorkerState::RUNNING &&
-      !DmtcpWorker::exitInProgress()) {
-    fprintf(stderr, "DMTCP INTERNAL ERROR: %s:%d: %s\n"
-            "       This process is not in RUNNING state and yet this thread\n"
-            "       managed to acquire the wrapperExecutionLock.\n"
-            "       This should not be happening, something is wrong.\n",
-            __FILE__, __LINE__, __PRETTY_FUNCTION__);
-    _exit(1);
+  // Ignore locks if we are about to exit
+  /* XXX:
+   *   One of the user threads could have called exit(), setting the
+   *   exitInProgress flag, and changing the WorkerState to Unknown. Does it
+   *   matter if we ignore the lock?
+   */
+  if (DmtcpWorker::exitInProgress()) {
```

Comment:
Actually, adding a warning is not straightforward. :-)

There are two difficulties with that:
 1) There can be a scenario where a user thread called exit, when another thread was in the middle of a wrapper, after having acquired the wrapperExecution lock. I would argue that it is safe to return from the unlock function w/o having done anything because the kernel is going to kill the process anyway.
 2) Even in the most simple case, we need some way for determining that a client is calling the unlock function even though the wrapperExecution lock was never acquired.

Adding a meaningful warning message requires maintaining a non-trivial global state that addresses the above issues. Please let me know what you think.